### PR TITLE
Separating the StandardWellsDense to interface file and implementation file

### PR DESCRIPTION
**Pull request by GitPaean**
It is first step of refactoring the StandardWellsDense, eventually to fit multisegment well  and potentially more different types of wells into the well model based on DenseAD. 

This PR only separates the `StandardWellsDense.hpp` to be a `.hpp` file and a `_impl.hpp` file, which potentially will make the interface of class `StandardWellsDense` more clear. 

A few `const`s are added and also `class WellState` in the function template is removed, since it should not be required. 

Tested with Norne, residual history is not changed and no obvious performance change is observed.

### CODE REVIEW

**Comment by GitPaean** [comment]
I can squash all the commits to be one if necessary. I just created them in case something wrong happened and it make it easier to spot the potential errors introduced.

**Comment by andlaus** [comment]
since this only affects wells stuff, and I'm quite incompetent in this regard, I'll have to refer this to @totto82 once again...

**Comment by andlaus** [comment]
(actually, I have an opinion about the `_impl.hpp` files: I do not like them because they tend to complicate things considerably and I don't really see any advantage of specifying the implementation directly behind the prototype. If @totto82 likes it better that way, I won't veto this PR, though.)

**Comment by GitPaean** [comment]
To me, the advantage is that the interface and structure of the class is much easier to tell by looking at the `.hpp` file. For this case, the `.hpp` file is 353 lines, which is about 5 screens on my display, comparing to about 30 screens in the old way.  For the developers that know all the details, it might not be a problem, while for people reading a new code, it is really challenging, because you have to keep searching/jumping to understand how things are designed and work.  

And also, even for developers, I think separating the interface from the implementation can be beneficial  since things are just clearer. 

Anyway, it is just my own opinion.

**Comment by andlaus** [comment]
> To me, the advantage is that the interface and structure of the class is much easier to tell by looking at the .hpp file.

kate supports code folding:

perf-n_linear1

and also, you can have a look at the doxygen output for the purpose you've mentioned.

on the other hand, constantly needing to switch buffers  buffers in emacs (which is my main editor) drives me crazy, i.e., it takes me *longer* to understand how a class works, in particular considering the fact that most APIs in opm tend to be ad-hoc; i.e., it is usually not possible to just look at the prototypes of a class' members and get a reasonable understanding about how they are connected.

> And also, even for developers, I think separating the interface from the implementation can be beneficial since things are just clearer.

In my experience this pattern just gets in the way when you want to change the interface of a class. (e.g. it is always a lot of fun to add or modify template parameters.)

> Anyway, it is just my own opinion.

same here ;)

**Comment by GitPaean** [comment]
> In my experience this pattern just gets in the way when you want to change the interface of a class. (e.g. it is always a lot of fun to add or modify template parameters.)

Maybe, while not for this case. Actually, the separation process helped to remove the unneeded `class WellState`. It causes a compilation error there.

**Comment by andlaus** [comment]
I cannot really see why this shouldn't also be the case for the "all-in-one" approach, but let's let @totto82 decide in the case of `StandardWellsDense`.

**Comment by atgeirr** [comment]
> I cannot really see why this shouldn't also be the case for the "all-in-one" approach, but let's let

I think he meant that it was *spotted* when splitting, not that it could not have removed it from the all-in-one header.

I am in favour of this change, for the record.

**Comment by atgeirr** [comment]
jenkins run norne please

**Comment by atgeirr** [comment]
I think that we might never all agree on the issue of splitting out the header, but I think it is best if code is readable as-is without special tools, even if those tools are rather easy to get or start using. Also, just as tools can make it easier to handle all-in-one-file (code collapsing), tools can also make it easier to handle split files (jumping between interface/implementation). Still, on GitHub for example there are only quite limited ways to manipulate this, and in general many people expect to be able to read the code to learn what it can do, which is in my opinion helped by a well-documented header. Not that the header here is documented... (yet, but it only moves code).

In conclusion, I'll merge this.

**Comment by andlaus** [comment]
I strongly recommend to let the people who work on this code the most decide. (for the wells stuff, these guys seem to be @totto82 and to some lesser extend @GitPaean. since I'm not one of these people, I did not raise a veto in this particular instance ;))

> (yet, but it only moves code).

weren't there also some (smallish) unrelated cleanups?

on a slightly related topic: I think we should start to think about providing up-to-date doxygen documentation for master on the website similar to how dune does it: https://www.dune-project.org/doxygen/master/

**Comment by GitPaean** [comment]
> weren't there also some (smallish) unrelated cleanups?

Yes.  There should not be any behavior change.  Adding `const`s should be unrelated.  And removing `WellState` kind of related/unrelated. It just causes compilation problem when keeping using it after separated. 

Thanks for your comments anyway.

**Comment by atgeirr** [comment]
> on a slightly related topic: I think we should start to think about providing up-to-date doxygen documentation for master on the website similar to how dune does it: https://www.dune-project.org/doxygen/master/

I agree, but it was a lot of hassle to get it right, wasn't it? Perhaps we could learn from how Dune is doing it. The Dune doxygen pages are indispensable for me, it would be good to have that for OPM, too.

**Comment by andlaus** [comment]
> I agree, but it was a lot of hassle to get it right, wasn't it?

in principle it is just a cronjob that runs doxygen once per night. As far as I can see, the main issue is integration of the result with the web site layout. On that regard I have to confess that I've got no experience of how to do this with wordpress, but for plain-PHP sites (*ehem* I'm looking at you old website!) it is relatively straight-forward. I think that layout integration should initially not be prioritized over to the contents of the documentation, though.
